### PR TITLE
[FFI] Fix the issue with the struct FF/DD in upcall on z/OS

### PR DESCRIPTION
**Pull request by ChengJin01**
[FFI] Fix the issue with the struct FF/DD in upcall on z/OS

The changes resolve the issue with the register mapping for the
struct FF/DD or its variants in upcall on z/OS by categorizing
them into ALL_SP/ALL_DP placed on FPRs while other non-complex
type are categorized into AGGREGATE_OTHER placed on GPRs.

Fixes: #19952

Signed-off-by: ChengJin01 <jincheng@ca.ibm.com>

### CODE REVIEW

**Comment by ChengJin01** [comment]
Review: @zl-wang (the code related to the encoding signature), @tajila (the FFI code in java)
FYI: @dchopra001, @r30shah, @joransiu, @pshipton

**Comment by ChengJin01** [comment]
The PR has been verified with FFI specific test suites in OpenJ9.

File: runtime/vm/LayoutFFITypeHelpers.hpp [hunk 1]
```diff
@@ -420,11 +458,23 @@ class LayoutFFITypeHelpers
...
+		 * the non-complex types are categorized into AGGREGATE_OTHER placed on GPRs.
+		 *
+		 * e.g.
+		 * 1) struct {float, union {float, float}} belong to ALL_SP while
+		 *    struct {double, union {double, double}} belongs to ALL_DP.
+		 * 2) struct {float a[2]} ([2:F] in terms of the layout string)
+		 *    and struct {double b[2]} ([2:D] in terms of the layout string)
+		 *    belong to AGGREGATE_OTHER placed on GPRs.
+		 */
+		if (isAllSP && isStructWithFFOrDD(initSig, sizeInByte, J9_FFI_UPCALL_STRUCT_FF_SIZE)) {
```
Comment (zl-wang) [inline runtime/vm/LayoutFFITypeHelpers.hpp:473]: 1) as far as i know, literally only [float, float] and [double, double] are considered as complex on zOS. any other (e.g. [float, {float, float}]) is not. so, the current impl is not exactly right;
2) for zOS, i didn't see [2:F] is differentiated from [F,F] either;

I will suggest a correct implementation as follows in `parseStruct`:  under cases of digits (for array) and [{ (for nested struct/union), adding
```
if (isZos()) {
   *isAllDP = false;
   *isAllSP = false;
}
```
since you are aiming for ALL_SP/DP to represent complex(es) on zOS.

@r30shah or @joransiu  is my understanding of complex types on zOS correct?

**Review by zl-wang — CHANGES_REQUESTED** [review]

**Comment by zl-wang** [comment]
also, is there any special treatment in zOS ABI for single-member struct/union?  something like:
```
struct {
   int me;
}
```
Hopefully no ... such that it is good as it is.

File: runtime/vm/LayoutFFITypeHelpers.hpp (hunk 1 above)
Comment (r30shah) [reply runtime/vm/LayoutFFITypeHelpers.hpp:473]: Hi @zl-wang , As @ChengJin01  as pointed in the comment, on z/OS ABI, a struct of size 8 bytes with two members , each is of same size (union of multiple floats only - The one that comment highlights), is considered complex. Similar for Double.

**Comment by r30shah** [comment]
> also, is there any special treatment in zOS ABI for single-member struct/union? something like:

If the passing parameter is union, it will regardless of member type, it will be passed in GPR. For the single element struct, it will be in GPR as well (whether it is Floating point or not).

File: runtime/vm/LayoutFFITypeHelpers.hpp (hunk 1 above)
Comment (zl-wang) [reply runtime/vm/LayoutFFITypeHelpers.hpp:473]: thanks for the clarification. 
```
typedef struct { float a; float b}  Definitely_Complex
typedef union { Definitely_Complex a; float b;} Not_Complex_Anymore
typedef struct { Definitely_Complex a;} Not_Complex_Either
```
I guessed these are true according to @r30shah clarification, such that the implementation still needs to handle these and the array case correctly.

Comment (zl-wang) [reply runtime/vm/LayoutFFITypeHelpers.hpp:473]: on the other hand, 
`typedef struct { struct {float a;} b; float c; }  This_is_a_Complex
`
currently, this is handled correctly.

**Review by zl-wang — CHANGES_REQUESTED** [review]

**Comment by zl-wang** [comment]
> If the passing parameter is union, it will regardless of member type, it will be passed in GPR. For the single element struct, it will be in GPR as well (whether it is Floating point or not).

for union treated as you described, is it limited to single-member and no-more-than-8-byte? or, you meant any union?
how about nested single-member struct?  something like:  `struct { struct {int a; int b;} c; }`  similar questions as well.

**Comment by r30shah** [comment]
> for union treated as you described, is it limited to single-member and no-more-than-8-byte? or, you meant any union?
how about nested single-member struct? something like: struct { struct {int a; int b;} c; } similar questions as well.

According to LE Vendor reference [1], for argument passing, any aggregate type that is not union and contains exactly two floating point types of same size, is considered complex type. All other parameters  from the argument lists are passed in GPR1-3 if available (With exception of floating point type. 

So the case you shared, if we have nested single member struct, as the aggregate has only one member, it will be passed in available GPRs.
```
For the purposes of argument passing and function return
values, in every language, every aggregate that is (a) not a union, and (b) contains exactly two floatingpoint
types of the same size (4,8, or 16 bytes) is treated as a complex type.
```


[1]. https://www.ibm.com/docs/en/SSLTBW_2.5.0/pdf/ceev100_v2r5.pdf

**Comment by zl-wang** [comment]
ok ... that makes it clear.  the implementation only needs to tighten up on a few cases mentioned above, and we are good to go.

**Comment by ChengJin01** [comment]
@zl-wang, is there is anything left to be updated against the clarification by @r30shah? If not, we should be able to get this merged asap.

**Comment by zl-wang** [comment]
@ChengJin01   at least three categorization bugs I can see, mentioned previously:
1) `struct { float a[2]; }`     should not be complex (even though ALL_SP and 8-byte in size);
2) `union { float a[2];  float b}`  should not be complex (even though ALL_SP and 8-byte in size);
3) `struct { struct {float a; float b;} c;} `should not be complex either (even though ALL_SP and 8-byte in size);  (@r30shah  agree? i am not sure about this case)
while you are fixing these, keep in mind the few being-complex cases previously.

**Comment by ChengJin01** [comment]
> @ChengJin01 at least three categorization bugs I can see, mentioned previously:
> 
>     1. `struct { float a[2]; }`     should not be complex (even though ALL_SP and 8-byte in size);
> 
>     2. `union { float a[2];  float b}`  should not be complex (even though ALL_SP and 8-byte in size);
> 
>     3. `struct { struct {float a; float b;} c;} `should not be complex either (even though ALL_SP and 8-byte in size);  (@r30shah  agree? i am not sure about this case)
>        while you are fixing these, keep in mind the few being-complex cases mentioned previously.

@dchopra001, please help confirm whether these 3 cases above are covered in this PR against z/OS ABI as the fix is totally based on z/OS ABI.

**Comment by r30shah** [comment]
I agree with the cases @zl-wang shared, all those should be considered non complex.

**Comment by ChengJin01** [comment]
@dchopra001,

please help confirm whether these 3 cases above are covered in this PR against z/OS ABI as the fix is totally based on z/OS ABI (my understanding is that case 2 & 3 should be addressed as explained at https://github.com/eclipse-openj9/openj9/pull/20018#issuecomment-2302479915).

**Comment by ChengJin01** [comment]
@zl-wang & @r30shah,

Do the rules also apply to `double`?
e.g.
```
struct { double a[2]; } should not be complex (even though ALL_DP and 16-byte in size);
union { double a[2];  double b} should not be complex (even though ALL_DP and 16-byte in size);
struct { struct {double a; double b;}} should not be complex either (even though ALL_DP and 16-byte in size);
```
If so, we will need to unify the code to handle them altogether.

**Comment by r30shah** [comment]
@ChengJin01  Yes , rules include all floating point types, so doubles as well.

**Comment by ChengJin01** [comment]
I've updated the code to handle all the 3 cases above for float & double.

**Comment by zl-wang** [comment]
@ChengJin01  
> As for case 3 `struct { struct {float a; float b;} float c;}` 

you added float before c  ... making it invalid declaration.

`struct { struct {float a; float b;}}` is an invalid type declaration as well  ... you need a field name at least (I named it `c` previously in my case 3).  but i understood you are referring to the same "type" as I ... it is 8-byte nested struct ALL_SP. since it has only one member ... so, it is not a complex. current code did it right?

**Comment by ChengJin01** [comment]
> so, it is not a complex. current code did it right?

Yes, the latest changes already covers this case (a struct with a 8-byte nested struct `ALL_SP`) as that's my understanding of our clarification above.

File: runtime/vm/LayoutFFITypeHelpers.hpp [hunk 2]
```diff
@@ -399,6 +400,52 @@ class LayoutFFITypeHelpers
...
+		 * Note:
+		 * The following float/double related union/structs ('T' stands for float or double)
+		 * are non-complex even though they are ALL_SP/ALL_DP with 8/16 bytes in size:
+		 * 1) union {T a[2]; T b;} which is "{2:F}" or "{2:D}" in the signature string.
+		 * 2) struct {T a[2];} which is "[2:F]" or "[2:D]"  in the signature string.
+		 * 3) struct {struct {T a; T b;} c;} which is "[[FF]]" or "[[DD]]" in the signature string.
+		 */
+		return ('{' != structSig[0]) /* A union starts from '{'. */
+				&& ('[' != structSig[1]) /* A nested struct ("[[FF]]" or "[[DD]]" starts from "[[". */
+				&& (':' != structSig[2]) /* "[2:F]" or "[2:D]" contains ':' at index 2. */
```
Comment (zl-wang) [inline runtime/vm/LayoutFFITypeHelpers.hpp:428]: how about  `struct { float a[1]; float b[1];}`?  that should be a complex.  the code above makes it not-complex? 
I can understand we probably belaboured on this point too much.  i think it would be good to go, once this corner case is fixed.

**Review by zl-wang — CHANGES_REQUESTED** [review]
i can live with the position we come back to fix this corner case in the future. @TobiAjila what do you think?

File: runtime/vm/LayoutFFITypeHelpers.hpp (hunk 2 above)
Comment (ChengJin01) [reply runtime/vm/LayoutFFITypeHelpers.hpp:428]: > how about struct { float a[1]; float b[1];}? that should be a complex. 

@r30shah, is it correct against the z/OS ABI? If so, I will need to slightly modify the code there to cover the case.

Comment (ChengJin01) [reply runtime/vm/LayoutFFITypeHelpers.hpp:428]: @zl-wang, how about `struct { float a[1]; float b; }` or ` struct { float a; float b[1]; }`? Should they be treated as complex?

Comment (zl-wang) [reply runtime/vm/LayoutFFITypeHelpers.hpp:428]: yeah, i think so

Comment (r30shah) [reply runtime/vm/LayoutFFITypeHelpers.hpp:428]: I do agree with @zl-wang. Such pointed out cases  would be considered complex atleast as per the documentation. I would have double checked to ensure but I am on a vacation so can not get on to system to verify. I would appreciate if @dchopra001  can verify this and confirm.

Comment (ChengJin01) [reply runtime/vm/LayoutFFITypeHelpers.hpp]: I just updated the code to only exclude `struct { struct {float a; float b;} c;}` as non-complex while `struct { float a[1]; float b[1];}`, struct { float a[1]; float b; }` and `struct { float a; float b[1]; }` are still complex.

File: runtime/vm/LayoutFFITypeHelpers.hpp [hunk 3]
```diff
@@ -399,6 +400,56 @@ class LayoutFFITypeHelpers
...
+		 * 2) struct {struct {T a[1]; T b;} which is "[1:FF]" or "[1:DD]" in the signature string.
+		 * 3) struct {struct {T a; T b[1];} which is "[F1:F]" or "[D1:D]" in the signature string.
+		 * [2] The following float/double related union/structs are non-complex even though they are
+		 * ALL_SP/ALL_DP with 8/16 bytes in size:
+		 * 1) union {T a[2]; T b;} which is "{2:F}" or "{2:D}" in the signature string.
+		 * 2) struct {T a[2];} which is "[2:F]" or "[2:D]" in the signature string.
+		 * 3) struct {struct {T a; T b;} c;} which is "[[FF]]" or "[[DD]]" in the signature string.
+		 */
+		return ('{' != structSig[0]) /* A union starts from '{'. */
+				&& ('[' != structSig[1]) /* A nested struct ("[[FF]]" or "[[DD]]" starts from "[[". */
```
Comment (zl-wang) [inline runtime/vm/LayoutFFITypeHelpers.hpp:431]: this might not be the right fix.  for example, [[[F]]F] ... i.e.  struct { struct { struct { float a;} b; } c; float d;}  this is still a complex i think. i knew we are hammering all kinds of corner cases here.

**Review by zl-wang — CHANGES_REQUESTED** [review]

File: runtime/vm/LayoutFFITypeHelpers.hpp (hunk 3 above)
Comment (ChengJin01) [reply runtime/vm/LayoutFFITypeHelpers.hpp:431]: What about `struct { struct { struct { float a; float b;}}}` ? Is it a complex? There might be a bunch of variants to be addressed.

Comment (zl-wang) [reply runtime/vm/LayoutFFITypeHelpers.hpp:431]: no, that is not, since it has only one member.

Comment (ChengJin01) [reply runtime/vm/LayoutFFITypeHelpers.hpp]: I replaced it with the following code which should cover all these nested cases (with one element).
```
				&& !strstr(structSig, "[[FF]]") /* A struct nested by "[[FF]]" or its variant. */
				&& !strstr(structSig, "[[DD]]") /* A struct nested by "[[DD]]". or its variant. */
```

File: runtime/vm/LayoutFFITypeHelpers.hpp [hunk 4]
```diff
@@ -399,6 +400,57 @@ class LayoutFFITypeHelpers
...
+#if defined(J9ZOS390)
+		/* Check whether the specified signature is a union (starting from '{')
+		 * which must be excluded given a union is treated as a non-complex
+		 * type according to the z/OS ABI.
+		 *
+		 * Note:
+		 * [1] The following float/double related structs ('T' stands for float or double) are complex:
+		 * 1) struct {struct {T a[1]; T b[1];}} which is"[1:F1:F]" or "[1:D1:D]" in the signature string.
+		 * 2) struct {struct {T a[1]; T b;} which is "[1:FF]" or "[1:DD]" in the signature string.
+		 * 3) struct {struct {T a; T b[1];} which is "[F1:F]" or "[D1:D]" in the signature string.
```
Comment (zl-wang) [inline runtime/vm/LayoutFFITypeHelpers.hpp:423]: nitpick:  type descriptions don't match with signature strings which are correct though.  these are not nested structs.

File: runtime/vm/LayoutFFITypeHelpers.hpp [hunk 5]
```diff
@@ -399,6 +400,57 @@ class LayoutFFITypeHelpers
...
+		 * ALL_SP/ALL_DP with 8/16 bytes in size:
+		 * 1) union {T a[2]; T b;} which is "{2:F}" or "{2:D}" in the signature string.
+		 * 2) struct {T a[2];} which is "[2:F]" or "[2:D]" in the signature string.
+		 * 3) struct {struct {T a; T b;} c;} which is "[[FF]]" or "[[DD]]" in the signature string.
+		 */
+		return ((structSize == layoutSize) /* The struct must be 8-byte or 16-byte in size. */
+				&& '{' != structSig[0]) /* A union starts from '{'. */
+				&& !strstr(structSig, "[[FF]]") /* A struct nested by "[[FF]]" or its variant. */
+				&& !strstr(structSig, "[[DD]]") /* A struct nested by "[[DD]]" or its variant. */
+				&& !(('2' == structSig[1]) && (':' == structSig[2])); /* "[2:F]" or "[2:D]" contains "2:" from index 1. */
```
Comment (zl-wang) [inline runtime/vm/LayoutFFITypeHelpers.hpp:434]: i think these tests cannot cover all the positive and negative cases respectively ... these probably are only good for catching a narrow band of them. i am also wondering if here is the right place to catch them all (or to saturate the right testing). 

in my mind, the better place to do this and hopefully catch them all:  somewhere in `parseStruct` ... eventually looking for an outer  struct with two members, 8/16 byte in size,  and isAllSP/isAllDP.

**Review by zl-wang — CHANGES_REQUESTED** [review]

File: runtime/vm/LayoutFFITypeHelpers.hpp (hunk 5 above)
Comment (ChengJin01) [reply runtime/vm/LayoutFFITypeHelpers.hpp:482]: > i think these tests cannot cover all the positive and negative cases respectively 

There is no 100% guarantee that the fix covers all corner cases which totally depend upon how they are handled on z/OS against the z/OS ABI. But the fix should already handle what explained by @r30shah as above.

Comment (ChengJin01) [reply runtime/vm/LayoutFFITypeHelpers.hpp:482]: > the better place to do this and hopefully catch them all: somewhere in parseStruct ... eventually looking for an outer struct with two members, 8/16 byte in size, and isAllSP/isAllDP.

Technically it might be possible to do that in `parseStruct` plus `setByteCellforStruct` and other methods called by `parseStruct` but essentially there is no difference and even more complicated/tricky as compared to the current fix because we still have to reply on signature string index to determine what the signature string looks like to handle these corner cases. And there is no way to guarantee that the new solution cover all corner cases even though we can achieve that.

**Review by zl-wang — APPROVED** [review]
i approved it on the basis that it catches the most likely cases (being good enough) and we are running out of time.

**Comment by ChengJin01** [comment]
@tajila, please double-check approve it if no more concerns.

**Comment by ChengJin01** [comment]
Just fixed the code conflicts when rebasing the branch.

**Review by tajila — APPROVED** [review]

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk21

**Comment by tajila** [comment]
jenkins test sanity xlinux64 jdk22

**Comment by tajila** [comment]
jenkins test sanity xlinux jdk22

**Comment by ChengJin01** [comment]
@dchopra001, please help confirm whether the fix works good in OpenJ9 FFI test suites.

**Comment by dchopra001** [comment]
> @dchopra001, please help confirm whether the fix works good in OpenJ9 FFI test suites.

This change fixes the issues we observed in openj9 functional tests.

We still see some failures in openjdk tests. I can't say with 100% certainty yet if they are all unrelated to this issue as I'm still working through them. However they don't look related so far.

**Comment by tajila** [comment]
jenkins test sanity zlinux jdk22

**Comment by ChengJin01** [comment]
The test failures at https://openj9-jenkins.osuosl.org/job/Test_openjdk21_j9_sanity.functional_aarch64_linux_Personal/204/tapTestReport/ (jobs were not yet finished) had nothing to do with FFI related tests (passed).

File: runtime/vm/LayoutFFITypeHelpers.hpp (hunk 5 above)
Comment (keithc-ca) [reply runtime/vm/LayoutFFITypeHelpers.hpp:482]: I don't think `strstr()` is the right tool for this task.

Comment (ChengJin01) [reply runtime/vm/LayoutFFITypeHelpers.hpp:482]: > I don't think `strstr()` is the right tool for this task.

What's your proposal for the replacement of `strstr()`?

Comment (ChengJin01) [reply runtime/vm/LayoutFFITypeHelpers.hpp:482]: As talked to @keithc-ca offline to avoid confusion, the fix is the simplest way cover all corner cases we encountered for now.
